docs/design_notes/lift.md

```diff
@@ -0,0 +1,338 @@
+    The updated variables are stored in the original/unlifted scopes such that the mutation that happened inside the transformation survive beyond the transformation boundary.
+
+
+### Using pack example
+
+
+A minimal example of using pack:
+
+```
+from flax.core import lift
+from flax.core import Scope, init, apply, nn as core_nn
+
+def lift_id(fn, variables=True, rngs=True)
+  def wrapper(scope_fn, repack_fn, variable_groups, rng_groups, *args):
+    # normally we would first call into a JAX transform here...
+    scope = scope_fn(variable_groups, rng_groups)
+    y = fn(scope, *args)
+    out_variables = repack_fn(scope)
+    return y, out_variables
+
+  return lift.pack(wrapper, (variables,), (variables,), (rngs,))
+
+x = jnp.ones((3, 2))
+y, params = init(lift_id(core_nn.dense))(random.PRNGKey(0), x, 4)
+```
+
+Normallly users shouldn't have to 
+
+### Supported transformatiosn
+
+| Jax Transform | Supported in Linen? | Comments |
+|-|-|-|
+| vmap | ✅ |  |
+| scan | ✅ | Carry variables cannot be initialized inside the scan body. |
+| remat | ✅ |  |
+| jit | ✅ | Current implementation might cause unnecessary recompilation. |
+| pmap | ❌ |  |
+| grad | ❌ |  |
+| xmap | ❌ |  |
+| custom_vjp | ❌ |  |
+| custom_jvp | ❌ |  |
+
+
+### Linen examples
+
+Going back to our original example we can now use the lifted vmap:
+
+```
+class LinenVmap(nn.Module):
+  @nn.compact
+  def __call__(self, xs):
+    VmapMLP = nn.vmap(MLP, variable_axes={'params': 0}, split_rngs={'params': True}, in_axes=0)
+    return VmapMLP(name='mlp')(xs)
+
+variables = LinenVmap().init(random.PRNGKey(0), xs)
+print(jax.tree_map(jnp.shape, variables['params']))
+"""==>
+{
+    mlp: {
+        Dense_0: {
+            bias: (3, 4),
+            kernel: (3, 2, 4),
+        },
+        Dense_1: {
+            bias: (3, 1),
+            kernel: (3, 4, 1),
+        },
+    },
+}
+"""
+```
+
+Here we use `variable_axes={'params': 0}` to indicate that paramaters are vectorized rather than shared and `split_rngs={'params': True}` means each set of parameters is initialized independently.
+
+We can also extend the example with some inner state by adding a `BatchNorm` layer:
+
+```
+class StatefulMLP(nn.Module):
+  @nn.compact
+  def __call__(self, x):
+    h = nn.Dense(4, name='hidden')(x)
+    h = nn.BatchNorm(use_running_average=False, axis_name="batch")(h)
+    h = nn.relu(h)
+    return nn.Dense(1, name='out')(h)
+
+class LinenStatefulVmap(nn.Module):
+  @nn.compact
+  def __call__(self, xs):
+    VmapMLP = nn.vmap(StatefulMLP, variable_axes={'params': 0, 'batch_stats': 0}, split_rngs={'params': True}, in_axes=0)
+    return VmapMLP(name='mlp')(xs)
+variables = LinenStatefulVmap().init(random.PRNGKey(0), xs)
+```
+
+All we had to add to `nn.vmap` is `'batch_stats': 0` indicating the batch stats are vectorized rather than shared along the first axis.
```

Comment:
Nit:

```suggestion
All we had to add to `nn.vmap` is `'batch_stats': 0`, indicating that the batch stats are vectorized rather than shared along the first axis.
```